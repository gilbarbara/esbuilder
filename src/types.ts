export interface SharedOptions {
  externals: boolean;
  noBundle: boolean;
  noSourcemap: boolean;
  outDirectory: string;
  platform: 'node' | 'browser';
  target: string;
  watch: boolean;
}

export interface Arguments extends SharedOptions {
  cjs: boolean;
  esm: boolean;
  input: string;
}

export interface BuildOptions extends SharedOptions {
  entryPoints: string[];
}
