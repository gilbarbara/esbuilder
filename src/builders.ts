/* eslint-disable no-console */
import * as chalk from 'chalk';
import esbuild from 'esbuild';
import { nodeExternalsPlugin } from 'esbuild-node-externals';

interface BuildOptions {
  entryPoints: string[];
  externals: boolean;
  noBundle: boolean;
  noSourcemap: boolean;
  outDirectory: string;
  platform: 'node' | 'browser';
  target: string;
}

export async function buildCJS(options: BuildOptions) {
  const { entryPoints, externals, noBundle, noSourcemap, outDirectory, platform, target } = options;

  console.log(chalk.blue('Building CommonJS...'));

  try {
    const cjsBuild = await esbuild.build({
      bundle: !noBundle,
      entryPoints,
      format: 'cjs',
      metafile: true,
      outdir: outDirectory,
      platform,
      plugins: externals ? [] : [nodeExternalsPlugin()],
      sourcemap: !noSourcemap,
      target,
    });

    /* istanbul ignore next */
    if (cjsBuild.warnings.length > 0) {
      console.error(cjsBuild.warnings);
    }

    const analysis = await esbuild.analyzeMetafile(cjsBuild.metafile);

    console.log(analysis);
    console.log(chalk.green('✓ CommonJS build complete.'));
  } catch (error) /* istanbul ignore next */ {
    console.error(error);
    process.exit(1);
  }
}

export async function buildESM(options: BuildOptions) {
  const { entryPoints, externals, noBundle, noSourcemap, outDirectory, platform, target } = options;

  console.log(chalk.blue('Building ESM...'));

  try {
    const esmBuild = await esbuild.build({
      bundle: !noBundle,
      entryPoints,
      format: 'esm',
      metafile: true,
      outdir: outDirectory,
      outExtension: { '.js': '.mjs' },
      platform,
      plugins: externals ? [] : [nodeExternalsPlugin()],
      sourcemap: !noSourcemap,
      target,
    });

    /* istanbul ignore next */
    if (esmBuild.warnings.length > 0) {
      console.error(esmBuild.warnings);
    }

    const analysis = await esbuild.analyzeMetafile(esmBuild.metafile);

    console.log(analysis);
    console.log(chalk.green('✓ ESM build complete.'));
  } catch (error) /* istanbul ignore next */ {
    console.error(error);
    process.exit(1);
  }
}
