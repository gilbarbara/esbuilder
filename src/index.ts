#!/usr/bin/env node
/* eslint-disable no-console,unicorn/prefer-top-level-await */

import { buildCJS, buildESM } from './builders';
import { Arguments } from './types';

const flags = require('yargs')(process.argv.slice(2)).argv;

const {
  cjs = false,
  esm = false,
  externals = false,
  input = './src/index.ts',
  noBundle = false,
  noSourcemap = false,
  outDirectory = './dist',
  platform = 'node',
  target = 'node16',
  watch = false,
} = flags as Arguments;

const entryPoints = input.split(',').map(entry => entry.trim());

/* istanbul ignore next */
(async () => {
  if (!cjs && !esm) {
    console.log(`Builds the source code into CommonJS and/or ESM modules.
    
Targets:
  --cjs            Build CommonJS module
  --esm            Build ESM module

Build Options:
  --input          Input files (default: ./src/index.ts)
  --externals      Add external packages to the bundle
  --noBundle       Skip bundling dependencies
  --noSourcemap    Skip the sourcemap
  --outDirectory   Output directory (default: ./dist)
  --platform       The platform (default: node)
  --target         The target environment (default: node16)
  --watch          Enabling watch mode (default: false)
`);

    process.exit(0);
  }

  if (cjs) {
    await buildCJS({
      entryPoints,
      externals,
      noBundle,
      noSourcemap,
      outDirectory,
      platform,
      target,
      watch,
    });
  }

  if (esm) {
    await buildESM({
      entryPoints,
      externals,
      noBundle,
      noSourcemap,
      outDirectory,
      platform,
      target,
      watch,
    });
  }
})();
