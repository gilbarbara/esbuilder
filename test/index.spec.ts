/* eslint-disable no-console */
const mockExit = jest.spyOn(process, 'exit').mockImplementation(() => undefined as never);

describe('index', () => {
  beforeAll(() => {
    jest.spyOn(console, 'log').mockImplementation(() => undefined);
    import('../src/index');
  });

  it('should return the help', () => {
    // @ts-ignore
    expect(console.log.mock.calls[0][0]).toMatchSnapshot();
    expect(console.log).toHaveBeenCalledTimes(1);
    expect(mockExit).toHaveBeenCalledWith(0);
  });
});
