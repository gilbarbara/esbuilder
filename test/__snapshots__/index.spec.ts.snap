// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index should return the help 1`] = `
"Builds the source code into CommonJS and/or ESM modules.
    
Targets:
  --cjs            Build CommonJS module
  --esm            Build ESM module

Build Options:
  --input          Input files (default: ./src/index.ts)
  --externals      Add external packages to the bundle
  --noBundle       Skip bundling dependencies
  --noSourcemap    Skip the sourcemap
  --outDirectory   Output directory (default: ./dist)
  --platform       The platform (default: node)
  --target         The target environment (default: node14)
"
`;
